### PR TITLE
Trying two e_v: one physical on as numerical regularization

### PR DESCRIPTION
**Pull request by mauro3**
Seems to do the job except that h[2,:] is wrong:
image

### CODE REVIEW

File: run_it_diffeq_expl.jl [hunk 1]
```diff
@@ -104,36 +103,40 @@ function make_ode()
         vc     .=  h .* (0.91 .* H .- ϕ).^3
 
         dhdt   .= (Σ .* vo .- Γ .* vc)
-        dϕdt   .= (.- (diff(flux_x[:,2:end-1],dims=1) ./ dx .+ diff(flux_y[2:end-1,:],dims=2) ./ dy) .- inn(dhdt) .+ Λ) ./ e_v
+        div_q  .= (diff(flux_x[:,2:end-1],dims=1) ./ dx .+ diff(flux_y[2:end-1,:],dims=2) ./ dy)
+        dϕdt   .= (.- div_q .- inn(dhdt) .+ Λ) ./ e_v
+        if use_masscons_for_h
+            # note this is 100% identical (modulus floating point errors) to use_masscons_for_h==false...
+            inn(dhdt) .= .-e_v.*dϕdt .- div_q .+ Λ
```
Comment (mauro3) [inline run_it_diffeq_expl.jl:110]: This is the mass-cons scheme of B&P.  However, note that it is 100% the same as doing the ODE.  I'm not sure whether in their FV-scheme there is a difference.

File: run_it_diffeq_expl_two_e_v.jl [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        vo     .= (h .< 1.0) .* (1.0 .- h)
+        vc     .=  h .* (0.91 .* H .- ϕ).^3
+
+        dhdt   .= (Σ .* vo .- Γ .* vc)
+        div_q  .= (diff(flux_x[:,2:end-1],dims=1) ./ dx .+ diff(flux_y[2:end-1,:],dims=2) ./ dy)
+        dϕdt   .= (.- div_q .- inn(dhdt) .+ Λ) ./ (e_v .+ e_v_num)
+
+        # update taking only non-regularization e_v into account
+        if use_masscons_for_h
+            # NOTE: above two are identical (modulus floating point errors)
```
Comment (mauro3) [inline run_it_diffeq_expl_two_e_v.jl]: Same here: 100% identical (modulus floating point errors)

File: run_it_diffeq_expl.jl (hunk 1 above)
Comment (pohlan) [reply run_it_diffeq_expl.jl:110]: Here the `dϕdt` in L110 is the updated one, shouldn't it refer to the one from the previous time step, if it is a purely explicit scheme?

Comment (mauro3) [reply run_it_diffeq_expl.jl:110]: No.  Explicit means that it doesn't use information from the "future" (which can only be gained by solving non-lin eqs).

Comment (pohlan) [reply run_it_diffeq_expl.jl:110]: Ah sorry I did a thinking mistake, I was just a bit confused that `dhdt` and `dϕdt` are using each other but then I saw that `dhdt` is updated twice. Clear now, thanks.

**Comment by mauro3** [comment]
Ok, issue reported above is fixed by resetting dphidt=0 at the edge.  Now get
image

File: run_it_diffeq_expl_two_e_v.jl [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            flux_y .= .- d_eff[:,2:end] .* max.(dϕ_dy, 0.0) .- d_eff[:,1:end-1] .* min.(dϕ_dy, 0.0)
+            vo     .= (h .< 1.0) .* (1.0 .- h)
+            vc     .=  h .* (0.91 .* H .- ϕ).^3
+
+            dhdt   .= (Σ .* vo .- Γ .* vc)
+            div_q  .= (diff(flux_x[:,2:end-1],dims=1) ./ dx .+ diff(flux_y[2:end-1,:],dims=2) ./ dy)
+            dϕdt   .= (.- div_q .- inn(dhdt) .+ Λ) ./ (e_v .+ e_v_num)
+            ## This fixes the issue reported in:
+            ## https://github.com/pohlan/SimpleSheet.jl/pull/4#issue-1041245216
+            dϕdt[1,:] .= 0 # BCs
```
Comment (mauro3) [inline run_it_diffeq_expl_two_e_v.jl:113]: This is the fix for the issue reported in the first post.

**Comment by pohlan** [comment]
As I explained in https://github.com/pohlan/SimpleSheet.jl/pull/5#commitcomment-59505648, it is not sufficient to set `dϕdt[1,:] .= 0` at the Dirichlet boundary as the update of `h` still depends on `div_q` (when the B&P equation is used for `h`), therefore one needs boundary conditions for `h` as well.

I fixed this in 84a9604 (on main branch). Here is a comparison, running with `use_masscons_for_h=true` for 100 days:

1. With bc for `h`, calculated from `dhdt = Σ vo - Γ vc` equation:
https://github.com/pohlan/SimpleSheet.jl/blob/84a9604fd2bd38e24f4901af4a4abfcf69ca49af/run_it_diffeq_expl_two_e_v.jl#L66
https://github.com/pohlan/SimpleSheet.jl/blob/84a9604fd2bd38e24f4901af4a4abfcf69ca49af/run_it_diffeq_expl_two_e_v.jl#L94-L96

100d_h_bc_correct

2. when commenting out bc for h:

100d_no_h_bc

Note: In this case it also works for `h[2,:]=1`, but in https://github.com/pohlan/SimpleSheet.jl/pull/5#commitcomment-59510828 it didn't (reason unknown).

**Comment by mauro3** [comment]
I just had a look at this in my other PR #10, there I set `dphidt[1:2,:]=0` as index 1 doesn't "count" anyway.  See
- https://github.com/pohlan/SimpleSheet.jl/pull/10/files#diff-552628110bf462afe9740aec68e2610fb12348047c003d7dc55c8cb225bf12ccR131-R132
- https://github.com/pohlan/SimpleSheet.jl/pull/10/files#diff-552628110bf462afe9740aec68e2610fb12348047c003d7dc55c8cb225bf12ccR108-R109

**Comment by pohlan** [comment]
But does it work for `use_masscons_for_h=true`? It shouldn't since `div_q` is not known here
https://github.com/pohlan/SimpleSheet.jl/blob/788936e9b8e4d2a9dcffbcb4f480a599fe71bd3f/run_it_diffeq_impl-sundials.jl#L132-L134

**Comment by mauro3** [comment]
I haven't tried for masscons=true yet.  Will report once I do.

**Pull request merged**